src/borg/testsuite/repo_new.py

```diff
@@ -0,0 +1,352 @@
+    assert repository.flags(H(0), mask=0xFFFFFFFF) == 0x00000000
+    assert repository.flags(H(1), mask=0xFFFFFFFF) == 0x00000006
+    # test case that doesn't work with remote repositories
+
+
+def _assert_sparse(repository):
+    # The superseded 123456... PUT
+    assert repository.compact[0] == 41 + 8 + len(fchunk(b"123456789"))
+    # a COMMIT
+    assert repository.compact[1] == 9
+    # The DELETE issued by the superseding PUT (or issued directly)
+    assert repository.compact[2] == 41
+    repository._rebuild_sparse(0)
+    assert repository.compact[0] == 41 + 8 + len(fchunk(b"123456789"))  # 9 is chunk or commit?
+
+
+def test_sparse1(repository):
+    repository.put(H(0), fchunk(b"foo"))
+    repository.put(H(1), fchunk(b"123456789"))
+    repository.commit(compact=False)
+    repository.put(H(1), fchunk(b"bar"))
+    _assert_sparse(repository)
+
+
+def test_sparse2(repository):
+    repository.put(H(0), fchunk(b"foo"))
+    repository.put(H(1), fchunk(b"123456789"))
+    repository.commit(compact=False)
+    repository.delete(H(1))
+    _assert_sparse(repository)
+
+
+def test_sparse_delete(repository):
+    chunk0 = fchunk(b"1245")
+    repository.put(H(0), chunk0)
+    repository.delete(H(0))
+    repository.io._write_fd.sync()
+    # The on-line tracking works on a per-object basis...
+    assert repository.compact[0] == 41 + 8 + 41 + len(chunk0)
+    repository._rebuild_sparse(0)
+    # ...while _rebuild_sparse can mark whole segments as completely sparse (which then includes the segment magic)
+    assert repository.compact[0] == 41 + 8 + 41 + len(chunk0) + len(MAGIC)
+    repository.commit(compact=True)
+    assert 0 not in [segment for segment, _ in repository.io.segment_iterator()]
+
+
+def test_uncommitted_garbage(repository):
+    # uncommitted garbage should be no problem, it is cleaned up automatically.
+    # we just have to be careful with invalidation of cached FDs in LoggedIO.
+    repository.put(H(0), fchunk(b"foo"))
+    repository.commit(compact=False)
+    # write some crap to an uncommitted segment file
+    last_segment = repository.io.get_latest_segment()
+    with open(repository.io.segment_filename(last_segment + 1), "wb") as f:
+        f.write(MAGIC + b"crapcrapcrap")
+    repository.reopen()
+    # usually, opening the repo and starting a transaction should trigger a cleanup.
+    repository.put(H(0), fchunk(b"bar"))  # this may trigger compact_segments()
+    repository.commit(compact=True)
+    # the point here is that nothing blows up with an exception.
+
+
+def test_replay_of_missing_index(repository):
+    add_keys(repository)
+    for name in os.listdir(repository.path):
+        if name.startswith("index."):
+            os.unlink(os.path.join(repository.path, name))
+    repository.reopen()
+    assert len(repository) == 3
+    # assert repository.check() is True - repository._active_txn is not false, causing check() to fail.
+    # my reopen method in repository isn't working as intended
```

Comment:
`assert repository.check() is True` fails because `repository._active_txn` is not false, causing the check to fail. This is all due to my reopen method not working as intended. The original reopen method created a new repository object, which resets this variable to false. 

I went in loops for hours today trying to get this working correctly, going to sleep on it and try again in the morning.